### DISCUSSION TITLE
Running any :Copilot command freezes vim entirely on large files (1k+ lines)

### DISCUSSION THREAD

**Discussion by tankorsmash** [community/community#52132] [Copilot Conversations] [answered]
### Select Topic Area

Bug

### Body

Windows 10
copilot.vim 1.8.3
Vim 9.0.0425
copilot/dist/agent.js 1.73.0
Node.js 16.17.0

Copilot works wonderfully on small files but every time I run even `:Copilot status` on certain Elm files 100% of the time, it hangs forever and have to close gvim via Windows, since no keypresses have any apparent effect.

It doesn't appear to be an issue with file size, since `:Copilot status` runs fine on a 4k JS file.

I couldn't find any related issues, but I may have missed something!

Potentially related, if I open the Elm file and run `:Copilot` immediately, the following error prints. If I run `:Copilot` again right after, the freeze occurs. This error doesn't show up when invoking Copilot in insert mode

```
line 2: sourcing "C:\Users\username\.vim\plugged\copilot.vim\autoload\copilot\panel.vim"
finished sourcing C:\Users\username\.vim\plugged\copilot.vim\autoload\copilot\panel.vim
continuing in 10
line 4: sourcing "C:\Users\username\.vim\plugged\copilot.vim\autoload\copilot\doc.vim"
finished sourcing C:\Users\username\.vim\plugged\copilot.vim\autoload\copilot\doc.vim
continuing in copilot#panel#Open
                                Error detected while processing function copilot#Command[38]..function copilot#Command[30]..10[2]..copilot#panel#Open[12]..copilot#Request[2]..<SNR>32_AgentRequest[21]..<SNR>32_AgentNotify[1]..<SNR>32_Send:
line    3:
E631: ch_sendraw(): Write failed

```

This occurs even after disabling all other plugins (save for colorschemes.vim).

**Comment by tankorsmash** [comment]
Found this [vim groups issue ](https://groups.google.com/g/vim_dev/c/UNNulkqb60k)where `ch_sendraw()` fails on large calls. My 1k lines Elm file is 89kb, whereas the 4k line TypeScript file is 60kb.

Editing the TypeScript file to make it double the size (with a lot of copy and pasting), making it roughly 120kb in size then exibits the same behaviour the Elm file did, where after inserting text, `:Copilot status` then freezes.

**Comment by chrisguilbeau** [comment]
Same problem here, freezes 100% of the time on larger files when enabling or running :Copilot status
Windows 10
vim: 9.0.1943
copilot.vim: 1.10
node: 18.18.0

These files work fine with copilot on emacs.

**Comment by todd-stanley** [comment]
I'm also experiencing the same problem with freezing with larger files with running :Copilot status.  My observation is that I can open a larger file and if I right away execute :Copilot status, it will report "Enabled and online".  If I then go into insert mode and back to normal mode, even if I don't actually make any edit, any subsequent attempts to execute :Copilot status will freeze.

Windows 10
copilot.vim 1.18.0
Vim 9.1.0104
dist/agent.js 1.158.0
Node.js 20.11.0

Was originally running Vim 9.1 and Node.js 18.18.0, and I upgraded them to see if that changed anything and it made no difference.  I tried Neovim 0.9.5 and that seems to work just fine with the same files and otherwise the same setup, so switching to Neovim seems like a potential workaround though it would be nice to have this working with Vim too.

**Comment by tpope** [comment]
The good news is that `ch_sendraw()` is gone. It's now using Vim's native JSON-RPC support (`:help language-server-protocol`), which I would expect to be more reliable. Hopefully that's the last of `E631: ch_sendraw(): Write failed`.

The bad news is I can't reproduce this, on the latest version of Copilot.vim or the version from a year ago. I edit big files like autoload/fugitive.vim (280K) a lot, so it's not for lack of trying.

It might be worth checking `:Copilot log`, though if it's hanging, it's quite possible nothing will show up.

**Comment by todd-stanley** [comment]
Thanks for looking into it.

I cloned down vim-fugitive and tried editing fugitive.vim and hit the issue right away.  That would suggest it's something different about my setup.

I then eliminated my .vimrc, all my other configurations and plugins except of course Copilot, and I still hit the issue.

I'm not getting anything in `:Copilot log`.  It's either empty or Vim is locked up.  I did open `:Copilot log`, noted the .log file name in AppData\Local\Temp, then replicated the problem and found the log file was empty.

One common theme with everyone in this thread reporting this issue does seem to be Windows 10.  The only thing I can think of is that is perhaps unusual with my setup is that my `$HOME` is `H:\` which is a mapped network drive.  Unfortunately that is managed by my IT department so it's not so easily changed on this computer.  But that does mean that Copilot is at `H:\vimfiles\pack\github\start\copilot.vim`.  I did also setup and try a secondary PC managed by my IT department, also with Windows 10 and encountered the same problem.  NeoVim seems to do things a bit differently in this regards.  `$HOME` is still `H:\`, but with NeoVim the configuration and plugins are at `C:\Users\<my log in name>\AppData\Local\nvim`.

**Reply by tpope** [reply]
> One common theme with everyone in this thread reporting this issue does seem to be Windows 10.

When I get a chance I'll try to reproduce there.

> The only thing I can think of is that is perhaps unusual with my setup is that my $HOME is H:\ which is a mapped network drive.

The plugin uses `$HOME` as the CWD for the child process. I wouldn't be surprised if that causes weirdness. Try this:

```vim
:let $HOME = 'C:\'
:echo expand('~')
:Copilot restart
```

If changing `$HOME` fixes it, then I can look into changing the default CWD.

If it doesn't fix it, it could be the fact that the plugin is _installed_ on a network drive. It would be worth exploring if you could get it to run when installed on a local drive, and if that fixes the problem. You can do `:set rtp^=C:\Users\myuser\vim-copilot` for example and install it to that directory.

**Reply by todd-stanley** [reply]
Thanks for the suggestions.  Unfortunately, none of them seem to help.  Changing the location of `$HOME` and restarting Copilot resulted in the same issue.  

I tried the suggestion with `:set rtp` and pointed it to a location on my C:\ drive, with no luck (though that's a handy trick... I might employ that anyway to move my configuration off the network drive in any case 😄).  I also simply tried copying the plugin to `C:\Program Files\Vim\vim91\pack\start\copilot.vim` since that's something I do have the ability to do, but no luck with that either.

Another random thought I had was to try the x86 build of Vim as an "official" x64 Windows build on vim.org is a relatively new thing, but that didn't help either.

**Comment by tankorsmash** [comment]
Thanks for looking into this too! My home is just in C:/ though, no network
mapping. Am on Windows 10 as well.

**Comment by tpope** [comment, marked as answer]
Fixed :crossed_fingers: in Copilot.vim 1.21.0.

**Reply by todd-stanley** [reply]
I just updated and did a quick check and I'm no longer able to replicate.  Woohoo!  👍 

copilot.vim 1.21.0
Vim 9.1.0104
GitHub Copilot Language Server 1.166.0
Node.js 20.11.0
Windows (Windows 10)

**Reply by tankorsmash** [reply]
Thank you @tpope!

**Reply by chrisguilbeau** [reply]
Worked for me too! Thank you @tpope!

**Comment by ettaboyle** [comment]
FYI @tpope I am going to reopen this discussion as it is solved and it may help other users.
`closed` discussions are not shown in default search and often missed by users.
